apps/recovery-relay/lib/wallets/ERC20/index.ts

```diff
@@ -1,73 +1,118 @@
-/* eslint-disable prefer-destructuring */
-import { Input } from '@fireblocks/wallet-derivation';
-import { Contract, Interface, Transaction, ethers } from 'ethers';
-import { AccountData, TxPayload, RawSignature } from '../types';
+import { Contract, ethers, JsonRpcProvider, parseUnits } from 'ethers';
+import { AccountData } from '../types';
 import { ConnectedWallet } from '../ConnectedWallet';
-import { Ethereum } from '../EVM/ETH';
+import { Input, EVMWallet as EVMBase } from '@fireblocks/wallet-derivation';
 import { erc20Abi } from './erc20.abi';
-import { transferAbi } from './transfer.abi';
-import BigNumber from 'bignumber.js';
+import { WalletClasses } from '..';
+import { EVM } from '../EVM';
 
-export class ERC20 extends Ethereum implements ConnectedWallet {
-  private contract: Contract;
+export class ERC20 extends EVMBase implements ConnectedWallet {
+  protected provider: JsonRpcProvider | undefined;
+  protected backendWallet: EVM | undefined;
+  public rpcURL: string | undefined;
+  public contract: Contract | undefined;
+  public tokenAddress: string | undefined;
+  public decimals: number | undefined;
+  public toAddress: string | undefined;
+  private normalizingFactor: string | bigint | undefined;
 
-  constructor(input: Input, tokenAddress: string) {
+  constructor(private input: Input) {
     super(input);
+  }
+
+  public setRPCUrl(url: string) {
+    this.backendWallet?.setRPCUrl(url);
+    //@ts-ignore
+    this.provider = this.backendWallet?.provider;
+  }
+
+  public setTokenAddress(address: string) {
+    this.tokenAddress = address;
+  }
+
+  public setNativeAsset(nativeAsset: String) {
+    this.backendWallet = new WalletClasses[nativeAsset as keyof typeof WalletClasses]({ ...this.input, assetId: nativeAsset });
+  }
+
+  public init() {
+    if (!this.tokenAddress) {
+      this.relayLogger.error(`ERC20 Token address unavailable: ${this.assetId}`);
+      throw new Error(`ERC20 Token address unavailable: ${this.assetId}`);
+    }
+
+    this.contract = new ethers.Contract(this.tokenAddress, erc20Abi, this.provider);
+
+    if (!this.contract) {
+      this.relayLogger.error(`ERC20 Token contract is undefined`);
+      throw new Error(`ERC20 Token contract is undefined`);
+    }
+  }
+
+  public setDecimals(decimals: number) {
+    this.decimals = decimals;
+    this.normalizingFactor = (10 ** decimals).toString();
+  }
 
-    this.contract = new ethers.Contract(tokenAddress, erc20Abi);
+  public setToAddress(toAddress: string) {
+    this.toAddress = toAddress;
   }
 
   public async getBalance(): Promise<number> {
-    this.weiBalance = await this.contract.balanceOf(this.address);
-    return parseFloat(parseFloat(ethers.formatEther(this.weiBalance)).toFixed(2));
+    const weiBalance: bigint = await this.contract?.balanceOf(this.address);
+    return Number(weiBalance / BigInt(this.normalizingFactor!));
```

Comment:
This might cause issues for very large balances
Better to keep it as BigInt and convert to hex string or to a normal string (base 10) when needed